---

## Section 1 - Data loading

let's load our data from the dataset provided.

In [14]:
import pandas as pd

In [15]:
df = pd.read_csv('GFCdata.csv')
df.head()

Company  Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft  2023       211915.0     72361.0      411976.0   
1  Microsoft  2022       198270.0     72738.0      364840.0   
2  Microsoft  2021       168088.0     61271.0      333779.0   
3      Tesla  2023        96773.0     14974.0      106618.0   
4      Tesla  2022        81462.0     12587.0       82338.0   

   Total Liabilities  Operating Activities  
0           205753.0               87582.0  
1           198298.0               89035.0  
2           191791.0               76740.0  
3            43009.0               13256.0  
4            36440.0               14724.0

---

## Section 2 - Data cleaning

Now that we have our dataset successfully loaded, we need to ensure that the data is clean. Data cleaning can be a very intense task, so for this exercise, we will focus just on ensuring that the correct datatypes are present for each column, and if not, correcting them.

We can use the `.info()` method to look at data types.

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Company               9 non-null      object 
 1   Year                  9 non-null      int64  
 2   Total Revenue         9 non-null      float64
 3   Net Income            9 non-null      float64
 4   Total Assets          9 non-null      float64
 5   Total Liabilities     9 non-null      float64
 6   Operating Activities  9 non-null      float64
dtypes: float64(5), int64(1), object(1)
memory usage: 632.0+ bytes


In [17]:
df.describe()

Year  Total Revenue    Net Income   Total Assets  \
count     9.000000       9.000000      9.000000       9.000000   
mean   2022.000000  217084.555556  59005.888889  268669.111111   
std       0.866025  133913.789489  38231.680758  140780.645366   
min    2021.000000   53823.000000   5644.000000   62131.000000   
25%    2021.000000   96773.000000  14974.000000  106618.000000   
50%    2022.000000  198270.000000  72361.000000  351002.000000   
75%    2023.000000  365817.000000  94680.000000  352755.000000   
max    2023.000000  394328.000000  99803.000000  411976.000000   

       Total Liabilities  Operating Activities  
count           9.000000              9.000000  
mean       125623.333333          69951.777778  
std         71667.954073          44632.990091  
min         30548.000000          11497.000000  
25%         43009.000000          14724.000000  
50%        145308.000000          87582.000000  
75%        191791.000000         104038.000000  
max        205753.000000         122151.000000

---

## Section 3 - Analyzing trends with pandas:

Use pandas to calculate year-over-year changes for each financial metric. You can do this by creating new columns in your DataFrame that represent the percentage change from one year to the next.

In [18]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [19]:
df

Company  Year  Total Revenue  Net Income  Total Assets  \
0  Microsoft  2023       211915.0     72361.0      411976.0   
1  Microsoft  2022       198270.0     72738.0      364840.0   
2  Microsoft  2021       168088.0     61271.0      333779.0   
3      Tesla  2023        96773.0     14974.0      106618.0   
4      Tesla  2022        81462.0     12587.0       82338.0   
5      Tesla  2021        53823.0      5644.0       62131.0   
6      Apple  2023       383285.0     96995.0      352583.0   
7      Apple  2022       394328.0     99803.0      352755.0   
8      Apple  2021       365817.0     94680.0      351002.0   

   Total Liabilities  Operating Activities  Revenue Growth (%)  \
0           205753.0               87582.0                 NaN   
1           198298.0               89035.0           -6.438902   
2           191791.0               76740.0          -15.222676   
3            43009.0               13256.0                 NaN   
4            36440.0               14724.0          -15.821562   
5            30548.0               11497.0          -33.928703   
6           145308.0              110543.0                 NaN   
7           153982.0              122151.0            2.881146   
8           125481.0              104038.0           -7.230275   

   Net Income Growth (%)  
0                    NaN  
1               0.520999  
2             -15.764800  
3                    NaN  
4             -15.940964  
5             -55.160086  
6                    NaN  
7               2.894995  
8              -5.133112

## Summary

## For Microsoft:

**Revenue Growth:**

2021 to 2022: -15.22%

2022 to 2023: NaN (not available)

**Net Income Growth:**

2021 to 2022: -15.76%

2022 to 2023: NaN (not available)

## For Tesla:

**Revenue Growth:**

2021 to 2022: -33.93%

2022 to 2023: NaN (not available)

**Net Income Growth:**

2021 to 2022: -55.16%

2022 to 2023: NaN (not available)

## For Apple:

**Revenue Growth:**

2021 to 2022: -7.23%

2022 to 2023: NaN (not available)

**Net Income Growth:**

2021 to 2022: -5.13%

2022 to 2023: NaN (not available)


The revenue and net income growth rates for 2022 to 2023 are not available for any of the companies.

In [20]:

# Assuming you have a DataFrame named 'df' with columns 'Company', 'Total Revenue', 'Net Income', and 'Year'
grouped_by_company_max = df.groupby(['Company']).agg({'Total Revenue': 'max', 'Net Income': 'max'})

# Get the corresponding year for maximum Total Revenue and Net Income
max_year = df.loc[df.groupby('Company')['Total Revenue'].idxmax(), ['Company', 'Year']]
grouped_by_company_max['Year_Max_Revenue'] = max_year.set_index('Company')['Year']

max_year = df.loc[df.groupby('Company')['Net Income'].idxmax(), ['Company', 'Year']]
grouped_by_company_max['Year_Max_Net_Income'] = max_year.set_index('Company')['Year']

print(grouped_by_company_max)


           Total Revenue  Net Income  Year_Max_Revenue  Year_Max_Net_Income
Company                                                                    
Apple           394328.0     99803.0              2022                 2022
Microsoft       211915.0     72738.0              2023                 2022
Tesla            96773.0     14974.0              2023                 2023


## Based on the provided data, we can analyze the revenue growth and net income growth for each company:


## Revenue Growth:

Apple experienced max growth in total revenue from an unspecified amount to $394,328 million in 2022.

Microsoft also saw max growth in total revenue from an unspecified amount to $211,915 million in 2023.

Tesla demonstrated max growth in total revenue from an unspecified amount to $96,773 million in 2023.

## Net Income Growth:

Apple showed max growth in net income from an unspecified amount to $99,803 million in 2022.

Microsoft exhibited max growth in net income from an unspecified amount to $72,738 million in 2022.

Tesla demonstrated  maxgrowth in net income from an unspecified amount to $14,974 million in 2023.

Overall, all three companies experienced growth in both total revenue and net income, indicating positive performance during the specified periods.







In [21]:

# Assuming you have a DataFrame named 'df' with columns 'Company', 'Total Revenue', 'Net Income', and 'Year'
grouped_by_company_min = df.groupby(['Company']).agg({'Total Revenue': 'min', 'Net Income': 'min'})

# Get the corresponding year for maximum Total Revenue and Net Income
max_year = df.loc[df.groupby('Company')['Total Revenue'].idxmax(), ['Company', 'Year']]
grouped_by_company_min['Year_Min_Revenue'] = max_year.set_index('Company')['Year']

max_year = df.loc[df.groupby('Company')['Net Income'].idxmax(), ['Company', 'Year']]
grouped_by_company_min['Year_Min_Net_Income'] = max_year.set_index('Company')['Year']

print(grouped_by_company_min)


           Total Revenue  Net Income  Year_Min_Revenue  Year_Min_Net_Income
Company                                                                    
Apple           365817.0     94680.0              2022                 2022
Microsoft       168088.0     61271.0              2023                 2022
Tesla            53823.0      5644.0              2023                 2023



## Based on the provided data, we can analyze the revenue growth and net income growth for each company:


## Revenue Growth:

Apple experienced min in total revenue from $365,817 million in 2022 to an unspecified amount.

Microsoft demonstrated min in total revenue from $168,088 million in 2023 to an unspecified amount.

Tesla showed min in total revenue from $53,823 million in 2023 to an unspecified amount.

## Net Income Growth:

Apple showed min in net income from $94,680 million in 2022 to an unspecified amount.

Microsoft exhibited min in net income from $61,271 million in 2022 to an unspecified amount.

Tesla demonstrated min in net income from $5,644 million in 2023 to an unspecified amount.

